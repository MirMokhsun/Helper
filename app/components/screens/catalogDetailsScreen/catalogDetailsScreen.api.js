/* eslint-disable eqeqeq */
/* eslint-disable no-plusplus, no-continue */
export const removeFromCurrentCards = (newCurrentCards, id) => {
    for (let i = 0; i < newCurrentCards.length; ++i) {
        if (newCurrentCards[i].key === id) {
            newCurrentCards.splice(i, 1);
            break;
        }
    }
    return newCurrentCards;
}

export const copyCurrentMapWithoutOne = (currentMap, id) => {
    const newMap = {};
    for (const key in currentMap) {
        if (key == id) continue;
        newMap[key] = currentMap[key]
    }
    return newMap;
}

export const getAllNames = (object) => {
    const result = [];
    if (object) {
        for (const key in object) {
            result.push(object[key].Name)
        }
    }
    return result;
}

export const makeObjectWithManeKeys = (object) => {
    const result = {};
    if (object) {
        for (const key in object) {
            if (object[key].recommendations && object[key].recommendations.summary) {
                result[object[key].security] = object[key].recommendations.summary;
            }
        }
    }
    return result;
}