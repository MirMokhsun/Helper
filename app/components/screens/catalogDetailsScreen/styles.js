import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
    scroll: {
        paddingTop: 20,
        backgroundColor: '#F5F5F7',

    },
    wrapper: {
        flex: 1,
        width: "100%",
        height: "100%",
        backgroundColor: '#F5F5F7',
    },
})