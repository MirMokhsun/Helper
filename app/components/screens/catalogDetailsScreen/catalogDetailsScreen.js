/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { View, FlatList } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { setCardItem } from '../detailCardScreen/redux/cardItemAction';
import { styles } from './styles';
import { setCurrentCatalogScreen, setFunctionNewQuotationMap } from './redux/currentScreenAction';
import { renderCards } from '../../../api/renderCards/renderCards';
import { localization } from '../../../common/localization';
import FreeServer from '../../../api/processingWithServer/getfreesrver';
import { showToast } from '../../commonComponents/contentWrapper/redux/showPopUpAction';
import { removeFromCurrentCards, copyCurrentMapWithoutOne, getAllNames, makeObjectWithManeKeys } from './catalogDetailsScreen.api';
import { IndicatorApi } from '../../../api/indicatorApi/indicatorApi';
import SearchToolBar from '../tabNavigatorScreens/tabBarScreen/toolbarInTabbar/searchToolBar/searchToolBar';

export class CatalogDetailsScreen extends Component {

    static navigationOptions = () => {
        return {
            header: null,
        }
    };

    constructor(props) {
        super(props)
        this.freeServer = new FreeServer();
        this.IndicatorApi = new IndicatorApi;
        this.state = {
            currentMap: {},
            currentCards: [],
            recomendationsForCards: {},
            searchCardText: '',
            // titleName: '',
            // screenName: '',
        }
        const { screenName, favoritesObj, titleName } = this.props.navigation.state.params;
        this.favoritesObj = favoritesObj;
        this.screenName = screenName;
        this.titleName = titleName;
    }

    componentDidMount = async () => {
        try {
            const { navigation, language, webSocketClient, getQuotationMap, setFunctionNewQuotationMap } = this.props;
            const { state: { routeName, params: { titleName } } } = navigation;
            let currentCards = [];
            currentCards = this.filterCards(this.props, '');
            this.setState({ currentCards, titleName });
            setFunctionNewQuotationMap(this.setNewQuotationMap);
            if (webSocketClient) {
                webSocketClient.webSocketClient.createWebSocketConnection();
            } else {
                this.freeServer.getFreeServer();
            }
            navigation.setParams({ title: localization[language]['Catalog'] });

            const allNames = getAllNames(getQuotationMap()[routeName]);
            await this.IndicatorApi.getTechAnalysis(allNames, ['DAY'])
                .then((responce) => {
                    if (responce.techAnalysis) {
                        const recomendations = makeObjectWithManeKeys(responce.techAnalysis);
                        this.setState({
                            recomendationsForCards: recomendations,
                        })
                    }
                });
        } catch (error) {

        }
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        const { navigation } = nextProps;
        const { state: { params: { titleName } } } = navigation;
        this.setState({ titleName });
    }

    UNSAFE_componentWillUnmount() {
        const { setCurrentCatalogScreen, setFunctionNewQuotationMap, webSocketClient } = this.props;
        if (webSocketClient) {
            webSocketClient.webSocketClient.closeWebSocketConnection();
        }
        setCurrentCatalogScreen('');
        setFunctionNewQuotationMap(null);
    };

    setTextRenderCards = (text) => {
        this.setState({ searchCardText: text });
    }

    filterCards = (props, searchCardText) => {
        const { getQuotationMap,
            language,
            setCardItem,
            navigation,
            setQuotationFavorites,
            showToast,
        } = props;
        const { state: { routeName, params: { titleName, screenName } } } = navigation;
        const { currentCards, recomendationsForCards } = this.state;
        let newCurrentCards = [];
        const { favoritesObj, removeFromCatalogDetailScreenCard } = this;
        newCurrentCards = renderCards(
            getQuotationMap()[routeName],
            currentCards,
            language,
            setCardItem,
            navigation,
            recomendationsForCards,
            favoritesObj,
            // titleName,
            this.screenName,
            setQuotationFavorites,
            removeFromCatalogDetailScreenCard,
            getQuotationMap,
            showToast,
            searchCardText,
            this.titleName
        );
        return newCurrentCards;
    }

    removeFromCatalogDetailScreenCard = (id) => {
        const { currentCards, currentMap } = this.state;
        const newCurrentCards = removeFromCurrentCards(currentCards, id);
        const newCurrentMap = copyCurrentMapWithoutOne(currentMap, id);
        this.setState({ currentCards: newCurrentCards, currentMap: newCurrentMap });
    }

    setNewQuotationMap = (data) => {
        const { currentMap, currentCards, recomendationsForCards, searchCardText } = this.state;
        const { language, setCardItem, navigation, setQuotationFavorites, showToast, getQuotationMap } = this.props;
        const { state: { params: { titleName, screenName } } } = navigation;
        const newMap = { ...currentMap, ...data };
        const newCurrentCards = renderCards(
            newMap,
            currentCards,
            language,
            setCardItem,
            navigation,
            recomendationsForCards,
            this.favoritesObj,
            // titleName,
            this.screenName,
            setQuotationFavorites,
            this.removeFromCatalogDetailScreenCard,
            getQuotationMap,
            showToast,
            searchCardText,
            this.titleName,
        );
        if (data) {
            this.setState({
                currentMap: newMap,
                currentCards: newCurrentCards
            });
        }
    }

    render() {
        const { currentCards } = this.state;
        return (
            <View style={styles.wrapper}>
                <SearchToolBar title={this.titleName} setTextRenderCards={this.setTextRenderCards} />
                <FlatList data={currentCards} renderItem={({ item }) => item} />
            </View>
        )
    }
}

const mapStateToProps = state => ({
    getQuotationMap: state.functionAddQuotationMap.getQuotationMap,
    setQuotationFavorites: state.functionAddQuotationMap.setQuotationFavorites,
    language: state.currentLanguage.language,
    webSocketClient: state.webSocketClient.webSocketClient,
    filterCatalogText: state.searchCatalog.filterCatalogText,
});

const mapDispatchToProps = dispatch => bindActionCreators({
    setCardItem,
    setCurrentCatalogScreen,
    setFunctionNewQuotationMap,
    showToast,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CatalogDetailsScreen);