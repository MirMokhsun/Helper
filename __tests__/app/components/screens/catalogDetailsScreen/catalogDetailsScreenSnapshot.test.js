import 'react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { CatalogDetailsScreen } from '../../../../../app/components/screens/catalogDetailsScreen/catalogDetailsScreen';

describe('CatalogDetailsScreen', () => {
    test('create snapshot CatalogDetailsScreen', () => {
        const renderer = new ShallowRenderer()
        let toolbarSnapshot = renderer.render(<CatalogDetailsScreen
        setCardItem={() => { }}
        setFunctionNewQuotationMap={() => { }}
        getQuotationMap={() => { return { 1: {} } }}
        setCurrentCatalogScreen={() => { }}
        webSocketClient={{ webSocketClient: () => { }, }}
        navigation={{ navigate: () => { }, setParams: () => { }, state: { params: 'test' } }}
        screenName={'1'}
        language={'ru'}
        />);
        expect(toolbarSnapshot).toMatchSnapshot();
    });
});
