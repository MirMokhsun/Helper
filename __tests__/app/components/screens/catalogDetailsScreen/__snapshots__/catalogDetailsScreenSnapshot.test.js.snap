// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogDetailsScreen create snapshot CatalogDetailsScreen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5F5F7",
      "flex": 1,
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Memo(Connect(SearchToolBar))
    setTextRenderCards={[Function]}
  />
  <FlatList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
